README.md
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit)](https://github.com/pre-commit/pre-commit)
[![ExpressJS](https://img.shields.io/badge/ExpressJS-blue?logo=ExpressJS)](https://nodejs.org/en)
[![NodeJS](https://img.shields.io/badge/nodejs-V18-brightgreen?logo=nodejs)](https://expressjs.com/en/api.html)

# health-check

Nodejs Health Check API that responds with a JSON object based on the health status of all other backend dependencies.

# Descriptions

In this project, we define an express app that listens on port `3005`. The app has one endpoint `/health` with `GET` method.

The `backendDependencies` array holds the names and URLs of all the backend dependencies that we want to check for health status.

When a `GET request` is made to `/health`, the server iterates over all backend dependencies and makes a GET request to their /health endpoint. If the request is successful, the dependency is marked as `"OK"` in the healthStatus object, otherwise it is marked as `"Error"`.

After all APIs have been checked, the isHealthy variable is set to true if all APIs are "OK", otherwise it is set to false.

Finally, the server responds with a JSON object containing the `isHealthy` flag and an array of results for each backend API.

This API allows for GET query requests and responds with a JSON object based on the health status of all backend dependencies. It can be used to quickly check the status of all backend dependencies and take appropriate action if any of them are not functioning correctly.

# Aurora DB Health Check

The `dbConfig` object holds the configuration information for our Aurora DB.

When a `GET request` is made to `/db/health`, the server attempts to create a connection to the Aurora DB using the mysql2/promise library. If the connection is successful, the server runs a simple `SELECT 1` query to check the health of the database. If the query is successful, the server returns a JSON object with a status of "OK", otherwise it returns a JSON object with a status of `"Error"`.

# Run in local

```bash
git clone https://github.com/jayantapaul-18/health-check.git
cd health-check
npm install
node index.js
or,
npm install -g nodemon
nodemon index.js
```

## Running pre-commit checks

[pre-commit](https://pre-commit.com) installs git hooks configured in [.pre-commit-config.yaml](.pre-commit-config.yaml)

Install `pre-commit` and `commitizen` to use

```bash
brew install commitizen
brew install pre-commit

pre-commit install
pre-commit install --hook-type commit-msg
pre-commit run --all-files

git add .
git status
pre-commit run --all-files
cz c
git commit -m 'feat: health check api with backend api status response'
git push origin main --force
```

# Release - standard-version

Run the below command to create a new release `CHANGELOGS`

```bash
npm install --save-dev standard-version
npm install --save-dev commitizen
npx commitizen init cz-conventional-changelog --save-dev --save-exact
# Install commitlint cli and conventional config
npm install --save-dev @commitlint/config-conventional @commitlint/cli
# Install Husky - Husky improves commits and more
npm install husky --save-dev
# Activate hooks
npx husky install
# Add hook
npx husky add .husky/commit-msg 'npx --no-install commitlint --edit "$1"'
# run
npm run commit
npm run release
git push --follow-tags origin main && npm publish
```

# To Bypass hooks`

We can bypass pre-commit and commit-msg hooks using Git -n/--no-verify option:

```bash
git commit -m "my no commit" --no-verify
```

# Request & Response

```bash
GET http://localhost:3005/health
```

`Response`

```bash
{"isHealthy":false,"healthStatus":[{"name":"API 1","status":"Error:connect ECONNREFUSED 127.0.0.1:4000"},{"name":"API 2","status":"Error:connect ECONNREFUSED 127.0.0.1:5000"},{"name":"API 3","status":"Error:connect ECONNREFUSED 127.0.0.1:6000"},{"name":"Database","status":"Error:connect ECONNREFUSED 127.0.0.1:3306"},{"name":"backend","status":"OK"}]}

```

# Using Docker

```bash
❯ docker build -t health-check .
- Need to optimize - TO DO
```

# Code Structure

```bash
tree -L 1
```

```bash
├── CHANGELOG.md
├── CODEOWNERS
├── Dockerfile
├── LICENSE
├── README.md
├── REQUEST.http
├── app.log
├── commitlint.config.js
├── config.js
├── index.js
├── node_modules
├── package-lock.json
├── package.json
├── prepare.js
└── swagger.json
```

# LICENSE


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2023-04-25)


### Features

* captured  exit request ([1dbe1bb](https://github.com/jayantapaul-18/health-check/commit/1dbe1bb373a0d6b69cd1deba780a480a2f95ea2c))
* health check api with backend api status response ([a5984a8](https://github.com/jayantapaul-18/health-check/commit/a5984a86a8c89a8f05f8789cce9a079fbcab1b36))


index.js
const express = require('express');
const http = require('http');
const originalRequest = http.request;
const axios = require('axios');
const helmet = require("helmet");
const bodyParser = require('body-parser');
const pino = require('pino-http')();
const mysql = require('mysql2/promise');
const { networkInterfaces } = require('os');
const winston = require('winston');
const logger = winston.createLogger({
  transports: [
    new winston.transports.Console(),
    new winston.transports.File({ filename: 'app.log' })
  ]
});
// const swaggerUi = require('swagger-ui-express');
// const swaggerDocument = require('./swagger.json');

const nets = networkInterfaces();
const results = Object.create(null); // Or just '{}', an empty object

for (const name of Object.keys(nets)) {
    for (const net of nets[name]) {
        // Skip over non-IPv4 and internal (i.e. 127.0.0.1) addresses
        // 'IPv4' is in Node <= 17, from 18 it's a number 4 or 6
        const familyV4Value = typeof net.family === 'string' ? 'IPv4' : 4
        if (net.family === familyV4Value && !net.internal) {
            if (!results[name]) {
                results[name] = [];
            }
            results[name].push(net.address);
        }
    }
}
const localIP = results.en0[0]
console.log(localIP);

/* Configure Express App */
const app = express();
const port = process.env.PORT || 3005;
// app.use(pino);
app.use(helmet());
// parse application/x-www-form-urlencoded
app.use(bodyParser.urlencoded({ extended: false }))
// parse application/json
app.use(bodyParser.json())
//app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));

app.use((req, res, next) => {
  logger.info(`${req.method} ${req.url}`);
  next();
});

http.request = function (options, callback) {
  if (options.host == undefined) {
    logger.info(`${options.method} ${localIP}:${options.port}${options.path}`);
  } else {
    logger.info(`${options.method} ${options.host}:${options.port}${options.path}`);
  }
  return originalRequest.call(this, options, callback);
};
process.on('exit', (code) => {
  logger.info(`Exit code: ${code}`);
});

const dbConfig = {
    host: 'your-db-hostname',
    user: 'your-db-username',
    password: 'your-db-password',
    database: 'your-db-name',
    port: 3306
};

/* Configure Backend API Health Check  */
const backendDependencies = [
    { name: 'API 1', url: 'http://localhost:4000' },
    { name: 'API 2', url: 'http://localhost:5000' },
    { name: 'API 3', url: 'http://localhost:6000' },
    { name: 'Database', url: 'http://localhost:3306' },
    { name: 'backend', url: 'https://geniushubbd.com/node'}
];

// Health-API
app.get('/health', async (req, res) => {
  const healthStatus = [];

  for (const dependency of backendDependencies) {
    try {
        const response = await axios.get(`${dependency.url}`);
        const result = { name: dependency.name, status: "OK"};
        healthStatus.push(result);
    } catch (err) {
        healthStatus[dependency.name] = (err.message);
        const result = { name: dependency.name, status: 'Error:'+err.message };
        healthStatus.push(result);
    }
  }

const isHealthy = healthStatus.every(result => result.status === 'OK');

res.status(isHealthy ? 200 : 500).json({ isHealthy, healthStatus });
});

app.get('/db/health', async (req, res) => {
    try {
      const connection = await mysql.createConnection(dbConfig);
      await connection.query('SELECT 1');
      await connection.end();
      res.status(200).json({ status: 'OK' });
    } catch (err) {
      console.error(err);
      res.status(500).json({ status: 'Error' });
    }
  });

/* Configure Express Server */
app.listen(port, () => {
  console.log(`Health check API listening at http://localhost:${port}/health`);
});
